src/ui/pages/SwapForm/SlippageSettings/SlippageSettings.tsx

```diff
@@ -91,6 +96,74 @@ function SlippageWarning({ percentValue }: { percentValue: string }) {
   );
 }
 
+const INPUT_TEXT_KIND = 'body/accent';
+
+function PercentWidth({ onValue }: { onValue(width: number): void }) {
+  const [show, setShow] = useState(true);
+  const percentCharWidthRef = useRef<HTMLDivElement | null>(null);
+
+  useLayoutEffect(() => {
+    if (percentCharWidthRef.current) {
+      onValue(percentCharWidthRef.current.clientWidth);
+      setShow(false);
+    }
+  }, [onValue]);
+
+  if (!show) {
+    return null;
+  }
+
+  return (
+    <UIText
+      ref={percentCharWidthRef}
+      kind={INPUT_TEXT_KIND}
+      style={{ visibility: 'hidden', position: 'absolute' }}
+    >
+      %
+    </UIText>
+  );
+}
+
+const MAX_OVERLAY_TEXT_LENGTH = 8;
+
+function CustomValueOverlay({
+  isOptimal,
+  value,
+}: {
+  isOptimal: boolean;
+  value: string | null;
+}) {
+  if ((value?.length || 0) > MAX_OVERLAY_TEXT_LENGTH) {
+    return null;
+  }
+
+  return (
+    <div
+      style={{
+        position: 'absolute',
+        inset: '0 0 0 0',
+        pointerEvents: 'none',
+        display: 'flex',
+        justifyContent: 'center',
+        alignItems: 'center',
+      }}
+    >
+      {value != null ? (
+        <UIText
+          kind={INPUT_TEXT_KIND}
+          color={isOptimal ? undefined : 'var(--notice-500)'}
+        >
+          {value}%
```

Comment:
I think `value` text should be transparent
If we also set overflow: hidden to the parent div, we won't need to hide the whole component based on length